doublesymmetry/SwiftAudioEx
# Feature/queue improvements

QueueManager:
- Fixes https://github.com/doublesymmetry/react-native-track-player/issues/1691
- **(breaking)** Make `playWhenReady` optional defaulting to `willPlayWhenReady` in `QueuedAudioPlayer#load`, `QueuedAudioPlayer#add`, `QueuedAudioPlayer#jump` & `AudioPlayer#load`
- **(breaking)** stop `QueuedAudioPlayer#next` and `QueuedAudioPlayer#previous` from (wrongly) mutating `playWhenReady`
- avoid catch logic in `QueuedAudioPlayer#AVWrapperItemDidPlayToEndTime`,  `QueuedAudioPlayer#next()` & `QueuedAudioPlayer#previous()` by adding (optional) wrap parameter to `QueueManager#next` and `QueueManager#previous`
- in QueuedAudioPlayer#AVWrapperItemDidPlayToEndTime when repeat mode is track, we can be assured playWhenReady = true since there is no way `AVWrapperItemDidPlayToEndTime` could be called otherwise. Therefore just seek to the start of the track and play.
- introduce private `skip` helper to `QueueManager` for use in `QueueManager#next` and `QueueManager#previous`
- have currentItem become nil after `clearQueue()`, `removeItem(at index: currentIndex)`
- avoid `moveItem(` from removing and re-adding item, causing it to load again
- avoid throwing in removePreviousItems() and removeNextItems() without queue items
- for methods that need to have tracks in the queue to function, throw new `APError.QueueError.empty` error
- gather invalid index error throwing into `throwIfIndexInvalid(index: Int, name: String = "index”)`
- remove activating of first added item and leave it open for consumers of QueueManager to do so by using onReceivedFirstItem delegate
- throw APError.QueueError.noCurrentItem when calling skip without current item
- allow jumping to current item (which would restart the audio)
- introduce `delegate?.onCurrentItemChanged` to detect switching of current item
- **(breaking)** rename addItems and addItem to add

Event
- **(breaking)** replace queueIndex event with currentItem event with an index parameter

APError
- remove noNextWhenRepeatModeTrack which wasn’t thrown anywhere
- add `.empty` which is thrown by functions expecting the queue not to be empty
- add `.noCurrentItem` which is thrown by methods expecting a current item to be set

QueuedAudioPlayer
- load no longer throws
- move loading / resetting into to QueuedAudioPlayer#onCurrentItemChanged
- move automatic activation of first added item into QueuedAudioPlayer#onReceivedFirstItem

AVPlayerWarapper
- expose AVPlayerWarapper #playWhenReady
- **(breaking)** remove soft parameter of reset
- change player state to idle after reset

AudioPlayer
- hook up AudioPlayer#playWhenReady to AVPlayerWrapper#playWhenReady

QueueManagerTests
- reflect that queue.current stays `nil` when first track is added
- in current item tests, reflect that jump now allows jumping to currentitem
- reflect that removing of `currentItem` is now possible
- add test for removing the `currentItem` when it is the last in the list
- reflect that jumping to the current item no longer throws an error
- reflect that moving the current item no longer throws an error and the currentIndex is changed
- reflect that clearing the queue changes `currentIndex` to `-1`

QueuedAudioPlayerTests
- test that when removing current item, the player becomes idle
- Adjust event listening code to use `audioPlayer.event.currentItem`
-  Adjust tests that expect there to be a currentItem after something has been added, by inserting a context step that does a `jump(to:0)`
- test that calling next or previous without a currentItem throws

Example Player
- Only show play when paused
- Enable queue repeat mode
